File: spine-cpp/src/spine/Skin.cpp
#include <stdexcept>
#include <spine/Attachment.h>
#include <spine/Skin.h>
#include <spine/BaseSkeleton.h>
#include <spine/Slot.h>

namespace spine {

Skin::Skin (const std::string &name) :
				name(name) {
}

Skin::~Skin () {
	for (std::map<Key, Attachment*>::iterator iter = attachments.begin(); iter != attachments.end(); iter++)
		delete iter->second;
}

void Skin::addAttachment (int slotIndex, const std::string &name, Attachment *attachment) {
	if (!attachment) throw std::invalid_argument("attachment cannot be null.");
	Key key = {slotIndex, name};
	attachments[key] = attachment;
}

Attachment* Skin::getAttachment (int slotIndex, const std::string &name) {
	Key key = {slotIndex, name};
	if (attachments.find(key) != attachments.end()) return attachments[key];
	return 0;
}

void Skin::attachAll (BaseSkeleton *skeleton, Skin *oldSkin) {
	for (std::map<Key, Attachment*>::iterator iter = attachments.begin(); iter != attachments.end(); iter++) {
		const Key key = iter->first;
		Slot *slot = skeleton->slots[key.slotIndex];
		if (slot->attachment == iter->second) {
			Attachment *attachment = getAttachment(key.slotIndex, key.name);
			if (attachment) slot->setAttachment(attachment);
		}
	}
}

} /* namespace spine */

File: spine-corona/spine/Skin.lua

local Skin = {}
function Skin.new (name)
	if not name then error("name cannot be nil", 2) end
	
	local self = {
		name = name,
		attachments = {}
	}

	function self:addAttachment (slotIndex, name, attachment)
		if not name then error("name cannot be nil.", 2) end
		self.attachments[slotIndex .. ":" .. name] = { slotIndex, name, attachment }
	end

	function self:getAttachment (slotIndex, name)
		if not name then error("name cannot be nil.", 2) end
		local values = self.attachments[slotIndex .. ":" .. name]
		if not values then return nil end
		return values[3]
	end

	function self:findNamesForSlot (slotIndex)
		local names = {}
		for k,v in self.attachments do
			if v[1] == slotIndex then table.insert(names, v[2]) end
		end
	end

	function self:findAttachmentsForSlot (slotIndex)
		local attachments = {}
		for k,v in self.attachments do
			if v[1] == slotIndex then table.insert(attachments, v[3]) end
		end
	end

	return self
end
return Skin
